rest/data/project_settings.go

```diff
@@ -97,6 +97,97 @@ func disableStartingSettings(p *model.ProjectRef) {
 	p.CommitQueue.Enabled = utility.FalsePtr()
 }
 
+// PromoteVarsToRepo moves variables from an attached project to its repo.
+// Promoted vars are removed from the project as part of this operation.
+// Variables whose names already appear in the repo settings will be overwritten.
+func PromoteVarsToRepo(projectId string, varNames []string, userId string) error {
+	project, err := model.GetProjectSettingsById(projectId, false)
+	if err != nil {
+		return errors.Wrapf(err, "getting project settings for project '%s'", projectId)
+	}
+	projectVars, err := model.FindOneProjectVars(projectId)
+	if err != nil {
+		return errors.Wrapf(err, "getting project variables for project '%s'", projectId)
+	}
+
+	repo, err := model.GetProjectSettingsById(project.ProjectRef.RepoRefId, true)
+	if err != nil {
+		return errors.Wrapf(err, "getting repo settings for repo '%s'", project.ProjectRef.RepoRefId)
+	}
+	repoVars, err := model.FindOneProjectVars(project.ProjectRef.RepoRefId)
+	if err != nil {
+		return errors.Wrapf(err, "getting repo variables for repo '%s'", project.ProjectRef.RepoRefId)
+	}
+
+	// Add each promoted variable to existing repo vars
+	apiRepoVars := &restModel.APIProjectVars{}
+	apiRepoVars.BuildFromService(*repoVars)
+	for _, varName := range varNames {
+		// Ignore nonexistent variables
+		if _, contains := projectVars.Vars[varName]; !contains {
+			continue
+		}
+		// Variables promoted from projects will overwrite matching repo variables
+		apiRepoVars.Vars[varName] = projectVars.Vars[varName]
+		if _, contains := projectVars.PrivateVars[varName]; contains {
+			apiRepoVars.PrivateVars[varName] = true
+		}
+		if _, contains := projectVars.AdminOnlyVars[varName]; contains {
+			apiRepoVars.AdminOnlyVars[varName] = true
+		}
+	}
+
+	if err = UpdateProjectVars(project.ProjectRef.RepoRefId, apiRepoVars, true); err != nil {
+		return errors.Wrapf(err, "updating project variables for project '%s'", projectId)
```

Comment:
`errors.Wrapf(... "updating repo ref variables for repo '%s'", project.ProjectRef.RepoRefId)`